### PR TITLE
Gui editor metadata

### PR DESCRIPTION
**Pull request by vinhui**
I added a section in the GUI editor's property tab for editing and viewing metadata of controls.
It only makes strings, numbers and booleans editable, but it will display objects as a JSON string.

It hides any metadata keys that start with an `_` or if the key is `guiEditor` (because all controls in the editor get that key).

A good snippet to test with is `#CITCWC#52` because it has some pre-existing metadata items.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14442/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14442/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14442/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14442/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14442/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14442/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14442/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14442/merge#BCU1XR#0

**Review by RaananW — COMMENTED** [review]
Love this!!!

Got a few questions. Currently it looks like this:

image

What I can see is that objects and arrays cannot be edited, and that primitives are converted to a string. But they are not converted back to their primitive value.

I like the way you allow editing keys directly, but this will make it hard to set the right value.

What we eventually did in the inspector is allow a textarea with the metadata, and allow the developer to edit this text (and format it to make it easier to edit). You can see that in the inspector:

image

**Comment by vinhui** [comment]
Primitives are converted to strings for the input part, but are then converted back to base types if applicable. This happens here https://github.com/vinhui/Babylon.js/blob/gui-editor-metadata/packages/tools/guiEditor/src/components/propertyTab/propertyGrids/gui/commonControlPropertyGridComponent.tsx#L228

The keys are not editable. I made them disabled text fields so that it's displayed with a fixed with (like a label), but also to indicate that they _are_ variable. The reason I made them non editable is that it gets weird if you loop over a dictionary to show all the input fields, but then also allow you to rename the key. This results in items jumping around, but also having to wait in code till the user has finished typing to apply the key rename.

The raw json editor is indeed also a nice way to go, but it's more "developery". Since I imagine the GUI editor to be used by other people besides developers, I'd prefer to make it simpler and less error-prone for them.

**Review by RaananW — DISMISSED** [review]
Convinced :-)

I am fine with the change. I'll ask @carolhmj to merge it if she is fine with it.

**Comment by RaananW** [comment]
Sorry - I have approved, but have a small request - 

image

Let's move it all the way down, as others (like the text block) have higher priority in the right panel.

**Comment by RaananW** [comment]
good question :-)

yeah, the architecture doesn't allow it to be added simply at the end, except for if it moves to the propertyTabComponent. Which i am not a big fan of, TBH.

Pinging @carolhmj, as she has more inside knowledge to the architecture there. We might not have a good solution here, unless we unify the element-specific rendering pipeline.

Oh, one other thing - the "+" doesn't work in a playground I am testing with.

**Comment by vinhui** [comment]
I think I see why. The `prompt()` javascript function shows the prompt dialog in the owner window. I'm not entirely sure why though. If I try to replicate this with a jsfiddle then I do get the promt in the correct window: https://jsfiddle.net/vtowha71/

**Comment by vinhui** [comment]
Pushed a fix for the prompt now showing in the popup

**Review by carolhmj — CHANGES_REQUESTED** [review]
Thanks for the contribution! Just a comment:

File: packages/tools/guiEditor/src/components/propertyTab/propertyGrids/gui/commonControlPropertyGridComponent.tsx [hunk 1]
```diff
@@ -734,6 +787,67 @@ export class CommonControlPropertyGridComponent extends React.Component<ICommonC
...
+                            while (controls.some((x) => keyName in x.metadata)) {
+                                num++;
+                                keyName = input + num;
+                            }
+                            this._addOrUpdateMetadata({ [keyName]: "" });
+                        }}
+                    />
+                </div>
+                {this._getCommonPropertyKeys(controls.map((x) => x.metadata)).map((metaKey) => {
+                    if (metaKey === "guiEditor" || metaKey.startsWith("_")) {
```
[the remark this answers is not in the document]
Comment (vinhui) [reply packages/tools/guiEditor/src/components/propertyTab/propertyGrids/gui/commonControlPropertyGridComponent.tsx]: No worries, it's easy to add more things

**Review by RaananW — APPROVED** [review]
LGTM! 
Just run `npm run format:fix`  and commit the formatted code :-)

**Review by carolhmj — APPROVED** [review]

**Pull request merged**